torchvision/csrc/ops/quantized/cpu/qnms_kernel.cpp

```diff
@@ -0,0 +1,124 @@
+#include <ATen/ATen.h>
+#include <ATen/native/quantized/affine_quantizer.h>
+#include <torch/library.h>
+
+namespace vision {
+namespace ops {
+
+namespace {
+
+template <typename scalar_t>
+at::Tensor qnms_kernel_impl(
+    const at::Tensor& dets,
+    const at::Tensor& scores,
+    double iou_threshold) {
+
+  TORCH_CHECK(!dets.is_cuda(), "dets must be a CPU tensor");
+  TORCH_CHECK(!scores.is_cuda(), "scores must be a CPU tensor");
+  TORCH_CHECK(
+      dets.scalar_type() == scores.scalar_type(),
+      "dets should have the same type as scores");
+
+  if (dets.numel() == 0)
+    return at::empty({0}, dets.options().dtype(at::kLong)); 
+
+  const auto ndets = dets.size(0);
+  const double dets_scale = dets.q_scale();
+
+
+  auto x1_t = dets.select(1, 0).contiguous();
+  auto y1_t = dets.select(1, 1).contiguous();
+  auto x2_t = dets.select(1, 2).contiguous();
+  auto y2_t = dets.select(1, 3).contiguous();
+  auto order_t = std::get<1>(scores.sort(0, /* descending=*/true));
+  at::Tensor suppressed_t = at::zeros({ndets}, dets.options().dtype(at::kByte));
+  at::Tensor keep_t = at::zeros({ndets}, dets.options().dtype(at::kLong));
+  at::Tensor areas_t = at::zeros({ndets}, dets.options().dtype(at::kFloat));
+
+  auto suppressed = suppressed_t.data_ptr<uint8_t>();
+  auto keep = keep_t.data_ptr<int64_t>();
+  auto order = order_t.data_ptr<int64_t>();
+  auto x1 = x1_t.data_ptr<scalar_t>();
+  auto y1 = y1_t.data_ptr<scalar_t>();
+  auto x2 = x2_t.data_ptr<scalar_t>();
+  auto y2 = y2_t.data_ptr<scalar_t>();
+  auto areas = areas_t.data_ptr<float>();
+
+  auto areas_a = areas_t.accessor<float, 1>();
+  for (int64_t i = 0; i < ndets; i++) {
+    areas_a[i] = dets_scale * dets_scale * (x2[i].val_ - x1[i].val_) * (y2[i].val_ - y1[i].val_);
```

Comment:
given that we only use areas in a relative way, we don't need to multiply them by `dets_scale ** 2`.
Additionally, this could even be kept as `int32_t` or some larger than `uint8_t` type, as we would only be considering the quantized area here if we don't consider the scale

Another thing to keep in mind is that if the boxes are malformed (i.e., `x2 < x1`) the operation will underflow and the results won't be compatible with the fp32 version anymore